### ISSUE TITLE
Revision file for create-project command

### ISSUE THREAD

**Issue by ro0NL** [composer/composer#4293]
Hi,

We currently use the `create-project` command to create new projects based on our template project. Before we forked the template project.

We are testing some update processes, i.e. keep the project in sync with the template project. With forking this was easy.. as we can add the forked remote to our project (`git checkout origin/master && git merge template/master`). However history-wise this was not ideal so we switched to `create-project`.

The update process using `create-project` should be done using a patch file with all changes from the revision when the project was created to the latest revision.

We currently create a TEMPLATE_REVISION file manually which references the commit hash from our template project, so we later can create a patch file with all later changes.

I would be nice if this revision file was touched automatically by composer.. any thoughts? Perhaps the whole process of creating/applying the patch file (i.e. updating) could be done by composer... :|

`composer update-project`

**Comment by alcohol** [comment]
Can you describe step by step what you exactly mean?

So you run `create-project`, which clones your template and installs all dependencies.

Then what?
What is this template revision file you are talking about? 
What needs to happen with it?

**Comment by ro0NL** [comment]
`create-project` clones the (template) repository but removes .the git afterwards, i.e. an export. This export is then committed to a new project repository.

We mostly use `dev-master` as the latest version. So after `create-project` we have an export from the template repository on a certain commit hash (i.e. HEAD of that moment). What i like to see is a reference file (e.g.. `.composer`) with the cloned commit hash of the (template) repository. This way we can keep track of all changes between the the commit hash stored in `.composer` and HEAD of the template repository later on.

I guess if using tagged versions this would be less a problem.. however you still have no reference which tag (i.e. commit hash) is cloned/exported on that moment.

**Comment by alcohol** [comment]
So, as I understand it, you want to clone your project and use it as a template, hence the `create-project`, but then you want to track all changes to be able to patch your original template? Why not just simply clone your template from the start then?

**Comment by ro0NL** [comment]
The other way around.. we clone our template (`create-project` that is) which becomes a new project/git repository and gets its own development lifecycle. The export of the template repository would be the initial commit of the project repository, ie.

```
cd /path/to/new/project
composer create-project -s dev vendor/template-project --repository=http://localhost/template-project.git
git init
git remote add -m master origin http://localhost/new-project.git # ie some customer project
git add -A .
git commit -m "Initial project created from template"
git push --set-upstream origin master
```

So we have 2 or more repositories now;
- template-project.git
 - new-project.git (based on template-project.git)
 - other-project.git ....
 - you'll get the point

Now the template project gets new commits (i.e. fixes for new projects)... these commits can/should or must be applied to the individual project repositories. However we have no reference in the project repository with commit hash that was initially exported from the template repository by `create-project`.

If `create-project` would touch a reference file, ie `.composer` with the repository URL+commit hash in it we can track changes between the hash in `.composer` and `HEAD` of the remote.

**Comment by mediamonks-robert** [comment]
I think my use case might be related, just ran into it this morning. I want to log the git commit hash to my projects, to do this added a "post-root-package-install" command and tried to use 

    $event->getComposer()->getPackage()->getDistReference()

but this returns null while I expected it to return the actual commit hash of the project.

To clarify, I see this when running composer create-project:

    Installing me/my-package (dev-develop 65b81dca93eb47764157651e5eba4398dbe395d3)

and I just want to save the "65b81dca93eb47764157651e5eba4398dbe395d3" part too, dev-develop is available with

    $event->getComposer()->getPackage()->getVersion()

without problems. If this is not related enough to the current issue I can create a new issue if needed.

**Comment by ro0NL** [comment]
That would be the Do-It-Yourself fix :') my suggestion is an extra option for `create-project` which handles this out-of-the-box, i.e. `create-project --track-referece` or something. That would touch the `.composer` file (or something) and stores the commit hash. If the repository URL is tracked too composer could even generate the patch file (i.e. diff between `tracked-repository-url@tracked-commit-hash` - `tracked-repositor-url@HEAD`

**Comment by stof** [comment]
@mediamonks-robert use ``getSourceReference`` when you install from source. You may not have a dist distribution for your package

**Comment by mediamonks-robert** [comment]
@stof Literally seconds before you posted your comment I found out about that function and had high hopes it would solve my issue, however it also returns null :(

**Comment by mediamonks-robert** [comment]
@ro0NL yes, that would be great. I was just doing something manually because I could not find an option for it. When that also failed I found this issue.

+1 on a "--track-reference" or even better: a "track-reference": true in config.

**Comment by ro0NL** [comment]
@Seldaek any thoughts on this? Imho, the update process for a project created via `create-project` is currently open to interpretation.

**Comment by Seldaek** [comment]
Given that create project is basically a glorified git clone I wonder if it
isn't easier to build an external tool that does the create project logic
but then also keeps track and manages updates.. I'm not entirely convinced
this belongs in composer itself.

**Comment by alcohol** [comment]
I personally do not really see or understand the use-case here. I have to agree with @Seldaek as to whether or not this is something Composer should be capable of dealing with (it is not IMHO).

**Comment by ro0NL** [comment]
It could... not sure it should. Your call ;-)

I guess for now we're going to create our own tooliing script, something like `~/bin/create-project /path/to/project`

This would wrap the `composer create-project` call and touches some `.revision` file afterwards. Later on we can add more scripts, something like `~/bin/create-update-patch /path/to/project`, etc.

You can close this issue if you like, however I still believe this could be part of composer, just as much `create-project` is part of composer. For me it currently feels like a missing workflow piece.

**Comment by alcohol** [comment]
I do not understand why you do not just clone your project directly?

If you want to be able to push back changes, that makes the most sense to me.

**Comment by stof** [comment]
@Seldaek I'm not even convinced that ``create-project`` actually belongs to Composer. So I totally agree about considering that making it more complex is out of the scope of Composer.

**Comment by mediamonks-robert** [comment]
Composer already contains create-project so it seems like a logical addition to add the revision to composer.lock or some other file. Composer already knows the revision because it is shown while installing, it's just not stored somewhere for someone to use in a script now.

Sure I can create my own script but it will be 95% the same functionality Composer can already do (clone, install dependencies, run scripts) with the only addition of storing a revision.

**Comment by ro0NL** [comment]
Projects are always cloned  for development. It's not about cloning some project repo nor the template repo. We need a workflow that enables us to track differences between the template repo and a project repo. I.e. all the commits in the template repo that are pushed after some project was created.

I couldnt describe it more clearly...

@Seldaek do you call for an industry standard tool that wraps composer into some project workflow tool? composerflow? :|

**Comment by alcohol** [comment]
Uh, so this sounds exactly like you just need to use `git clone` and then you can use `git diff` to track your changes between local and remote.

**Comment by ro0NL** [comment]
Ok.. i got your point now :+1:  i guess this is possible and maybe even good enough :/

However the history will not match as all the commits of the template repo are pushed as one, e.g. 'Initial project mported' into the project repo.. im not that much if a git guru to forsee how `git diff` handles this.. my guess a diff of the later revisions only, the inititisl idea, would be more practical.

Still... i like the idea of an automated workflow for projects...

**Comment by alcohol** [comment]
Unless your template repo only has one commit, I do not understand what you mean with "all the commits of the template repo as pushed as one"? When you `git clone` an existing repository (your template in this case), you get the full history of that repository.

**Comment by mediamonks-robert** [comment]
That's the difference with my use case, I want to start clean because the project I create has it's own repository, I'm not interested in pushing back changes. It would be really nice to have the revision of that version so in case of one way patches from the template project to the actual project.

Will do some digging in Composer source to see if I can find an easy fix to get the vcs revision in root package object as getFullPrettyVersion() only returns "dev-develop" instead of "dev-develop <hash>".

**Comment by ro0NL** [comment]
The complete history (read: source code) of the template repo is committed to a new project repo after `create-project` has ran... 

So the source after 100+ commits in the template repo are pushed as one commit, the initial project commit.

**Comment by mediamonks-robert** [comment]
Maybe a solution for me would be to add "--keep-vcs" to my install command, use git to get the current revision and then delete the .git folder manually.

**Comment by alcohol** [comment]
Yes but that is because `create-project` does not keep the history of your template (the project you cloned). If you manually clone your template, you will retain this history.

**Comment by alcohol** [comment]
@mediamonks-robert I think Composer does not know the revision because after the `ProjectInstaller` finishes, it changes into the directory you specified, and it is not until then that it actually instantiates an instance of `Composer` using `Factory::create()`. So it just reads from the generated `composer.json` then.

**Comment by mediamonks-robert** [comment]
It does print the version and revision when installing the root package and the root package object is available inside the "postRootPackageInstall" script so I would expect to also have the information it prints to be available in the src/Composer/Package/RootPackage.php class.

**Comment by ro0NL** [comment]
Got it. I think adding the template repo as a 2nd remote on the project repo, i.e. `template-origin` and then fetch it could get us some nice diffs. I.e. it will show us the commit hash we're looking for..

**Comment by alcohol** [comment]
@mediamonks-robert no that package is only available in the `create-project` command. See the following lines:

https://github.com/composer/composer/blob/2438105e11da443f72628fb9d4bf4e089d229c6c/src/Composer/Command/CreateProjectCommand.php#L280
https://github.com/composer/composer/blob/2438105e11da443f72628fb9d4bf4e089d229c6c/src/Composer/Command/CreateProjectCommand.php#L301

I do not see any way at the moment to access this reference at a later point elsewhere.

**Comment by mediamonks-robert** [comment]
I was able to fix my issue by using the advantage that the *post-root-package-install"* script is ran before the .git directory is deleted. I could read all details from git in my script and let composer clean up my .git folder afterwards without needing the *--keep-vcs* option. Maybe this can help others too :)

**Comment by ro0NL** [comment]
Working with 2 remotes solves my original problem (`origin` vs. `template-origin`). So closing it...

As i like the idea for some sort of project manager, composer based, i'll dive into this in my own time.
